### PR TITLE
feat: introduce list_quorum="auto" to prefer quorum drives

### PR DESCRIPTION
**Pull request by harshavardhana**
## Community Contribution License
All community contributions in this pull request are licensed to the project maintainers
under the terms of the [Apache 2 license] (https://www.apache.org/licenses/LICENSE-2.0). 
By creating this pull request I represent that I have the right to license the 
contributions to the project maintainers under the Apache 2 license.

## Description
feat: introduce list_quorum="auto" to prefer quorum drives

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [x] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

**Comment by harshavardhana** [comment]
Submitted as a draft for some initial feedback and discussions.

**Comment by klauspost** [comment]
Well, this doesn't really include any of the "hard" part. We should probably discuss that, since I don't really see any way to do that in a useful/scalable way.

**Comment by harshavardhana** [comment]
> Well, this doesn't really include any of the "hard" part. We should probably discuss that, since I don't really see any way to do that in a useful/scalable way.

The next change I will make will be bucket centric and then subsequently try to go all the way to 1 level prefix but I have to test it further.

File: cmd/metacache-set.go [hunk 1]
```diff
@@ -590,25 +590,73 @@ func (er *erasureObjects) streamMetadataParts(ctx context.Context, o listPathOpt
...
+
+	commonMutations := calcCommonCounter()
+	for i, info := range infos {
+		if (info.Metrics.TotalWrites + info.Metrics.TotalDeletes) >= commonMutations {
+			newDisks = append(newDisks, disks[i])
+		}
+	}
+
+	return newDisks
+}
```
Comment (Praveenrajmani) [inline cmd/metacache-set.go]: :clap: :ok_hand:

File: cmd/metacache-set.go [hunk 2]
```diff
@@ -590,25 +590,73 @@ func (er *erasureObjects) streamMetadataParts(ctx context.Context, o listPathOpt
...
 	case "optimal":
 		return (driveCount + 1) / 2
+	case "auto":
+		return -1
 	}
 	// defaults to 'strict'
 	return driveCount
 }
 
+func getQuorumDisks(disks []StorageAPI, infos []DiskInfo, readQuorum int) (newDisks []StorageAPI) {
```
Comment (Praveenrajmani) [inline cmd/metacache-set.go]: Question: "auto" list quorum will work only if we see some common mutations in the disks? ie, if the disks go down and come up randomly, we won't be able to get some common mutations on (total deletes + total writes), and we end up choosing all the disks for the listing. Correct?

**Review by Praveenrajmani — APPROVED** [review]

File: cmd/metacache-set.go (hunk 2 above)
Comment (harshavardhana) [reply cmd/metacache-set.go]: Idea is that eventually all the missed events will get healed. Very flaky systems we can't do much they will behave the same as what it is today.

Remember this is erasure set level what it means is that there are many other erasure sets that may be good so overall it shall benefit. If one or two sets don't qualify it isn't a big deal

Comment (harshavardhana) [reply cmd/metacache-set.go]: Also for now this feature is kept opt-in, default will be still strict.

**Comment by klauspost** [comment]
@Praveenrajmani My instincts says the same. I don't really see this working - but Harsha convinced me enough to "at least we can give it a try".

My overall impression is that it relies too much on state that cannot be kept consistent and cannot be resynchronized once out-of-sync. If we were a DB with a strict journal "yes" - but we went the scalable direction.

My main issue is that once it desyncs (and it will, just do a rolling restart) it will bring down performance unexpectedly. So if systems are relying on this to keep afloat they will have a rude awakening once it de-syncs.

For a read-only cluster also "yes", but then you might as well just reduce the listing disks once and for all. But let's get some data instead of speculating.

@harshavardhana Is there a way for us to extract the write/delete numbers? 

It would be interesting to see these after running this for a month in various prod environments. Even being able to see how they change over time would be very valuable for evaluating the feasibility.

**Comment by harshavardhana** [comment]
> @Praveenrajmani My instincts says the same. I don't really see this working - but Harsha convinced me enough to "at least we can give it a try".
> 
> My overall impression is that it relies too much on state that cannot be kept consistent and cannot be resynchronized once out-of-sync. If we were a DB with a strict journal "yes" - but we went the scalable direction.
> 
> My main issue is that once it desyncs (and it will, just do a rolling restart) it will bring down performance unexpectedly. So if systems are relying on this to keep afloat they will have a rude awakening once it de-syncs.
> 
> For a read-only cluster also "yes", but then you might as well just reduce the listing disks once and for all. But let's get some data instead of speculating.
> 
> @harshavardhana Is there a way for us to extract the write/delete numbers?
> 
> It would be interesting to see these after running this for a month in various prod environments. Even being able to see how they change over time would be very valuable for evaluating the feasibility.

Yes @klauspost that information is persisted to disk in xattrs so you can use standard tools like getfattr to get those values and you can even manually sync them to a quorum state via setfattr and then restart services.

I have these Go specific tools already, will put these into our containers so that they can be used in case they are ever needed. 

Almost all our customers use strict quorum already and they are chugging along. However for customers who have stable infrastructures and manage their clusters well. Have the scale of many disks and erasure sets this will work fine for the most part.

**Comment by harshavardhana** [comment]
PR is updated with 

- xattr tool added for any debugging that we might have to do
- Handled drive replacement healing situations. Thanks to @krisis for reminding me to do that.

**Comment by harshavardhana** [comment]
The fresh drive replacement scenario requires a more fundamental change, and it will not work correctly in this PR.

This PR can only handle disconnects, and missed events are healed; however, it can't correctly heal a fresh drive replacement scenario.

Also, the persistent nature of this can only be enabled with `export _MINIO_DRIVE_QUORUM=on` environment variable.

**Comment by harshavardhana** [comment]
PTAL @vadmeste @krisis - I have changed this to make it an opt-in feature with `_MINIO_DRIVE_QUORUM=on`

However, disk replacement and healing have not yet been done due to the underlying challenges of re-using the `local disk`.

**Comment by harshavardhana** [comment]
I will take a stab at it once more by tonight and see where I can land with it.

**Comment by harshavardhana** [comment]
To correctly know the right values per disk we need this PR to go in first

https://github.com/minio/minio/pull/18645

**Comment by harshavardhana** [comment]
To use this PR and test you need to use _MINIO_DRIVE_QUORUM=on (to enable this feature), it is a hidden variable until it emerges to be default after some field understanding.

while testing make sure to do that.

**Review by minio-trusted — APPROVED** [review]

**Pull request merged**